py/server/deephaven/experimental/iceberg.py

```diff
@@ -0,0 +1,251 @@
+_JTableIdentifier = jpy.get_type("org.apache.iceberg.catalog.TableIdentifier")
+_JSnapshot = jpy.get_type("org.apache.iceberg.Snapshot")
+
+
+class IcebergInstructions(JObjectWrapper):
+    """
+    This class specifies the instructions for reading an Iceberg table into Deephaven. These include column rename
+    instructions and table definitions, as well as special data instructions for loading data files from the cloud.
+    """
+
+    j_object_type = _JIcebergInstructions
+
+    def __init__(self,
+                 table_definition: Optional[Union[Dict[str, DType], List[Column]]] = None,
+                 data_instructions: Optional[s3.S3Instructions] = None,
+                 column_renames: Optional[Dict[str, str]] = None):
+        """
+        Initializes the instructions using the provided parameters.
+
+        Args:
+            table_definition (Optional[Union[Dict[str, DType], List[Column], None]]): the table definition; if omitted,
+                the definition is inferred from the Iceberg schema. Setting a definition guarantees the returned table
+                will have that definition. This is useful for specifying a subset of the Iceberg schema columns.
+            data_instructions (Optional[s3.S3Instructions]): Special instructions for reading data files, useful when
+                reading files from a non-local file system, like S3.
+            column_renames (Optional[Dict[str, str]]): A dictionary of old to new column names that will be renamed in
+                the output table.
+
+        Raises:
+            DHError: If unable to build the instructions object.
+        """
+
+        try:
+            builder = self.j_object_type.builder()
+
+            if table_definition is not None:
+                builder.tableDefinition(j_table_definition(table_definition))
+
+            if data_instructions is not None:
+                builder.dataInstructions(data_instructions.j_object)
+
+            if column_renames is not None:
+                for old_name, new_name in column_renames.items():
+                    builder.putColumnRenames(old_name, new_name)
+
+            self._j_object = builder.build()
+        except Exception as e:
+            raise DHError(e, "Failed to build Iceberg instructions") from e
+
+    @property
+    def j_object(self) -> jpy.JType:
+        return self._j_object
+
+
+class IcebergCatalogAdapter(JObjectWrapper):
+    """
+    This class provides an interface for interacting with Iceberg catalogs. It allows listing namespaces, tables and
+    snapshots, as well as reading Iceberg tables into Deephaven tables.
+    """
+    j_object_type = _JIcebergCatalogAdapter or type(None)
+
+    def __init__(self, j_object: _JIcebergCatalogAdapter):
+        self.j_catalog_adapter = j_object
+
+    def namespaces(self, namespace: Optional[str] = None) -> Table:
+        """
+        Returns the namespaces in the catalog as a Deephaven table.
```

Comment:
Is this returning just the namespaces at the specified level, or is it returning all the namespaces in the specified branch of the tree?

Reply:
At the specified level (or root if not specified).  I'll improve the doc string.